### PR TITLE
do some small optimization to ops

### PR DESCRIPTION
**Pull request by njzjz**
avoid concat or add in loops. Instead, append tensors to a list, and concat or accumulate_n after loops

### CODE REVIEW

**Comment by amcadmus** [comment]
Have you benchmarked these optimization? Do they help improving the efficiency?

**Comment by njzjz** [comment]
> Have you benchmarked these optimization? Do they help improving the efficiency?

I just benchmarked it. The answer is no😂

**Comment by njzjz** [comment]
Do some profiling here:
(1) `+` vs `accumulate_n`
`+` one by one has more ops than `accumulate_n` once.

image
`+`

image
`accumulate_n`

(2) `concat`
image

File: deepmd/descriptor/se_a.py [hunk 1]
```diff
@@ -797,12 +798,12 @@ def _filter(
                       bavg = bavg,
                       trainable = trainable,
                       suffix = "_"+str(type_i))
-                  if type_i == 0:
```
Comment (wanghan-iapcm) [inline deepmd/descriptor/se_a.py:800]: Did we have a bug here? if `type_i == 0` and `(type_input, type_i)  in self.exclude_types` we had ret accumulated.

**Comment by wanghan-iapcm** [comment]
@denghuilu Would the revised code faster on GPUs?

**Comment by njzjz** [comment]
I think one cannot see any difference if there are only one or two types. A system with at least 10 types should be tested.

**Comment by denghuilu** [comment]
> 

There is a slight performance penalty on V100 GPU with the water benchmark system:

**optimize-ops branch**
```

DEEPMD INFO    batch     100 training time 3.36 s, testing time 2.34 s
DEEPMD INFO    batch     200 training time 1.73 s, testing time 2.32 s
DEEPMD INFO    batch     300 training time 1.75 s, testing time 2.32 s
DEEPMD INFO    batch     400 training time 1.73 s, testing time 2.41 s
DEEPMD INFO    batch     500 training time 1.72 s, testing time 2.37 s
DEEPMD INFO    batch     600 training time 1.74 s, testing time 2.36 s
DEEPMD INFO    batch     700 training time 1.76 s, testing time 2.43 s
DEEPMD INFO    batch     800 training time 1.77 s, testing time 2.48 s
DEEPMD INFO    batch     900 training time 1.75 s, testing time 2.47 s
DEEPMD INFO    batch    1000 training time 1.72 s, testing time 2.41 s

```
**devel branch**
```
DEEPMD INFO    batch     100 training time 3.03 s, testing time 0.02 s
DEEPMD INFO    batch     200 training time 1.60 s, testing time 0.02 s
DEEPMD INFO    batch     300 training time 1.63 s, testing time 0.02 s
DEEPMD INFO    batch     400 training time 1.59 s, testing time 0.02 s
DEEPMD INFO    batch     500 training time 1.58 s, testing time 0.02 s
DEEPMD INFO    batch     600 training time 1.62 s, testing time 0.02 s
DEEPMD INFO    batch     700 training time 1.59 s, testing time 0.02 s
DEEPMD INFO    batch     800 training time 1.58 s, testing time 0.02 s
DEEPMD INFO    batch     900 training time 1.60 s, testing time 0.02 s
```
Maybe the GPU implementation did not use the stream parallelization.

**Comment by wanghan-iapcm** [comment]
> > 
> 
> There is a slight performance penalty on V100 GPU with the water benchmark system:
> 
> **optimize-ops branch**
> 
> ```
> 
> DEEPMD INFO    batch     100 training time 3.36 s, testing time 2.34 s
> DEEPMD INFO    batch     200 training time 1.73 s, testing time 2.32 s
> DEEPMD INFO    batch     300 training time 1.75 s, testing time 2.32 s
> DEEPMD INFO    batch     400 training time 1.73 s, testing time 2.41 s
> DEEPMD INFO    batch     500 training time 1.72 s, testing time 2.37 s
> DEEPMD INFO    batch     600 training time 1.74 s, testing time 2.36 s
> DEEPMD INFO    batch     700 training time 1.76 s, testing time 2.43 s
> DEEPMD INFO    batch     800 training time 1.77 s, testing time 2.48 s
> DEEPMD INFO    batch     900 training time 1.75 s, testing time 2.47 s
> DEEPMD INFO    batch    1000 training time 1.72 s, testing time 2.41 s
> ```
> 
> **devel branch**
> 
> ```
> DEEPMD INFO    batch     100 training time 3.03 s, testing time 0.02 s
> DEEPMD INFO    batch     200 training time 1.60 s, testing time 0.02 s
> DEEPMD INFO    batch     300 training time 1.63 s, testing time 0.02 s
> DEEPMD INFO    batch     400 training time 1.59 s, testing time 0.02 s
> DEEPMD INFO    batch     500 training time 1.58 s, testing time 0.02 s
> DEEPMD INFO    batch     600 training time 1.62 s, testing time 0.02 s
> DEEPMD INFO    batch     700 training time 1.59 s, testing time 0.02 s
> DEEPMD INFO    batch     800 training time 1.58 s, testing time 0.02 s
> DEEPMD INFO    batch     900 training time 1.60 s, testing time 0.02 s
> ```
> 
> Maybe the GPU implementation did not use the stream parallelization.

Why the testing time of optimize-ops is so long?

**Comment by njzjz** [comment]
> 
> Why the testing time of optimize-ops is so long?

It was fixed by #1419 -- this branch is behind devel.

**Comment by denghuilu** [comment]
It did have some benefits:
```
DEEPMD INFO    batch     200 training time 1.59 s, testing time 0.02 s
DEEPMD INFO    batch     300 training time 1.56 s, testing time 0.02 s
DEEPMD INFO    batch     400 training time 1.57 s, testing time 0.02 s
DEEPMD INFO    batch     500 training time 1.59 s, testing time 0.02 s
DEEPMD INFO    batch     600 training time 1.59 s, testing time 0.02 s
DEEPMD INFO    batch     700 training time 1.60 s, testing time 0.02 s
DEEPMD INFO    batch     800 training time 1.60 s, testing time 0.02 s
DEEPMD INFO    batch     900 training time 1.60 s, testing time 0.02 s
DEEPMD INFO    batch    1000 training time 1.57 s, testing time 0.02 s
```

**Review by wanghan-iapcm — APPROVED** [review]

**Pull request merged**